### PR TITLE
Clean up filter generation

### PR DESCRIPTION
**Pull request by Kijewski**

### CODE REVIEW

File: askama_derive/src/generator.rs [hunk 1]
```diff
@@ -1210,66 +1210,106 @@ impl<'a> Generator<'a> {
...
+        if args.len() != 1 {
+            return Err("unexpected argument(s) in `safe` filter".into());
+        }
+        buf.write("::askama::filters::yaml(");
+        buf.write(self.input.escaper);
+        buf.write(", ");
+        self._visit_args(buf, args)?;
+        buf.write(")?");
+        Ok(DisplayWrap::Wrapped)
+    }
```
Comment (GuillaumeGomez) [inline askama_derive/src/generator.rs:1261]: Wouldn't it be simpler to instead have:

```suggestion
    fn _visit_yaml_filter(
        &mut self,
        buf: &mut Buffer,
        args: &[Expr<'_>],
    ) -> Result<DisplayWrap, CompileError> {
        if cfg!(not(feature = "serde-yaml")) {
            return Err("the `yaml` filter requires the `serde-yaml` feature to be enabled".into());
        }
        if args.len() != 1 {
            return Err("unexpected argument(s) in `safe` filter".into());
        }
        buf.write("::askama::filters::yaml(");
        buf.write(self.input.escaper);
        buf.write(", ");
        self._visit_args(buf, args)?;
        buf.write(")?");
        Ok(DisplayWrap::Wrapped)
    }
```

?

File: askama_derive/src/generator.rs [hunk 2]
```diff
@@ -1210,66 +1210,106 @@ impl<'a> Generator<'a> {
...
+    #[cfg(not(feature = "serde-json"))]
+    fn _visit_json_filter(
+        &mut self,
+        _: &mut Buffer,
+        _: &[Expr<'_>],
+    ) -> Result<DisplayWrap, CompileError> {
+        Err("the `json` filter requires the `serde-json` feature to be enabled".into())
+    }
+
+    #[cfg(feature = "serde-json")]
```
Comment (GuillaumeGomez) [inline askama_derive/src/generator.rs:1222]: Same as below.

File: askama_derive/src/generator.rs (hunk 1 above)
Comment (Kijewski) [reply askama_derive/src/generator.rs]: Yes, that's much better!

File: askama_derive/src/generator.rs [hunk 3]
```diff
@@ -1210,10 +1210,34 @@ impl<'a> Generator<'a> {
         Ok(DisplayWrap::Wrapped)
     }
 
+    #[cfg(not(feature = "serde-json"))]
+    fn _visit_json_filter(
```
Comment (djc) [inline askama_derive/src/generator.rs:1209]: Style nit: please follow top-down ordering for this -- `_visit_json_filter()` should go below `visit_filter()`. (Same for the others.)

**Review by djc — APPROVED** [review]
This is great, thanks!

File: askama_derive/src/generator.rs (hunk 3 above)
Comment (Kijewski) [reply askama_derive/src/generator.rs]: Moved around.

**Comment by Kijewski** [comment]
`_visit_as_ref_filter()` should return `DisplayWrap::Unwrapped`, shouldn't it? E.g. `{{ "<script>"|as_ref }}`. @GuillaumeGomez

**Comment by djc** [comment]
Ping?

**Comment by djc** [comment]
Yes, definitely! That's actually a security risk...

**Comment by Kijewski** [comment]
Rebased.

**Review by GuillaumeGomez — APPROVED** [review]

**Pull request merged**